reflex-dom-core/src/Reflex/Dom/WebSocket/Query.hs

```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecursiveDo #-}
+module Reflex.Dom.WebSocket.Query (cropQueryT, runWebSocketQuery) where
+
+import Data.Default
+import Control.Monad.Fix
+import Data.Text (Text)
+import Data.Aeson
+import Reflex
+import Reflex.Dom.WebSocket
+import Foreign.JavaScript.TH
+import Data.Maybe
+import Language.Javascript.JSaddle.Types (MonadJSM)
+
+runWebSocketQuery :: (MonadJSM m, MonadJSM (Performable m), HasJSContext m, PostBuild t m, TriggerEvent t m, PerformEvent t m, MonadHold t m, Reflex t, ToJSON q, MonadFix m, Query q, FromJSON (QueryResult q), Additive q, Group q, Eq q)
+                  => QueryT t q m a
+                  -> Text -- ^ WebSocket url
+                  -> m a
+runWebSocketQuery app url = do
+  postBuild <- getPostBuild
+  rec ws <- jsonWebSocket url $ def { _webSocketConfig_send = pure <$> updatedRequest }
+      (a, request) <- cropQueryT app $ fromMaybe mempty <$> _webSocket_recv ws
+      let updatedRequest = leftmost [updated request, tag (current request) postBuild]
+  return a
+
+cropQueryT :: (Reflex t, MonadHold t m, MonadFix m, Query q, Additive q, Group q, Eq q)
```

Comment:
Not sure what to name this one? Contrary to `runQueryT`, this also handles cropping the result.
The original intention was to reuse this for xhr, but that will require something like `RequesterT` since the responses may come out of order. So maybe we should just drop this?